src/backend/backend.ts

```diff
@@ -15,39 +14,67 @@ import {
   Message,
   MessageFactory,
   MessageType,
+  browserDispatch,
   browserSubscribe,
 } from '../communication';
 
 import {send} from './indirect-connection';
 
 import {
-  Route,
   MainRoute,
-  defineLookupOperation,
   highlight,
   parseRoutes,
 } from './utils';
 
+import {serialize} from '../utils';
+
+import {MessageQueue} from '../structures';
+
 import {SimpleOptions} from '../options';
 
 declare const ng;
 declare const getAllAngularRootElements: () => Element[];
 declare const treeRenderOptions: SimpleOptions;
 
+/// For tree deltas that contain more changes than {@link deltaThreshold},
+/// we simply send the entire tree again instead of trying to patch it
+/// since it will be faster than trying to apply hundreds or thousands of
+/// changes to an existing tree.
+const deltaThreshold = 512;
+
+/// For large messages, we do not send them through the normal pipe (which
+/// is backend > content script > backround channel > frontend), we add them
+/// to this buffer and then send a {@link MessageType.Push} message that
+/// tells the frontend to read messages directly from this queue itself.
+/// This allows us to prevent very large messages containing tree data from
+/// being serialized and deserialized four times. Using this mechanism, they
+/// are serialized and deserialized a total of one times.
+const messageBuffer = new MessageQueue<Message<any>>();
+
 /// NOTE(cbond): We collect roots from all applications (mulit-app support)
 let previousTree: MutableTree;
 
+let previousCount: number;
+
 const updateTree = (roots: Array<DebugElement>) => {
   const showElements = treeRenderOptions.showElements;
 
-  const newTree = createTreeFromElements(roots, showElements);
+  const {tree, count} = createTreeFromElements(roots, showElements);
+
+  if (previousTree == null || Math.abs(previousCount - count) > deltaThreshold) {
+    messageBuffer.enqueue(MessageFactory.completeTree(tree));
+  }
+  else {
+    messageBuffer.enqueue(MessageFactory.treeDiff(previousTree.diff(tree)));
+  }
+
+  /// Send a message through the normal channels to indicate to the frontend
+  /// that messages are waiting for it in {@link messageBuffer}
+  send<void, void>(MessageFactory.push());
 
-  send<void, any>(
-    previousTree
-      ? MessageFactory.treeDiff(previousTree.diff(newTree))
-      : MessageFactory.completeTree(newTree));
+  previousTree = tree;
```

Comment:
In the future let's extract the variable assignment out of it. One way is to have a redux store on the backend that will manage this. This is completely out of scope though.